arch/arm/src/stm32h7/stm32h743xx_flash.c

```diff
@@ -769,6 +769,12 @@ ssize_t up_progmem_eraseblock(size_t block)
       return (ssize_t)ret;
     }
 
+  /* Flush and invalidate D-Cache for this block */
+
+#ifdef CONFIG_ARMV7M_DCACHE
+  up_flush_dcache(block_address, block_address + FLASH_SECTOR_SIZE);
```

Comment:
shouldn't this be a `up_invalidate_dcache()` call and called before `stm32h7_israngeerased()`?

Reply:
HI @pkarashchenko. The problem I face is a lack in synchronization between the sector erase inside `up_progmem_eraseblock()` and the following erase check in `stm32h7_israngeerased()`.

I see that once the sector is erased the cache flush the precious contents and once I check the memory status it appears as not clean, so in my opinion the flush of cache from previous operations must be done before the sector erase.

Regarding the use of `up_flush_cache()` in place of `up_invalidate_dcache()` is for using the most conservative Cache operation. `up_invalidate_dcache()`  marks the contents as invalid (basically, a delete operation). 

Info form: [level-1-cache-on-stm32f7-series-and-stm32h7-series-stmicroelectronics.pdf](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwik8O-7vfb6AhWIhv0HHcrTCt4QFnoECBcQAQ&url=https%3A%2F%2Fwww.st.com%2Fresource%2Fen%2Fapplication_note%2Fdm00272913-level-1-cache-on-stm32f7-series-and-stm32h7-series-stmicroelectronics.pdf&usg=AOvVaw22vWOqlHEuqOoGdM3hJ8sd)

What do you think?

Reply:
Let me take a closer look later today. In general  the logic that I use is next:
1. If memory is modified by CPU and then modified area is read by peripheral (or DMA or any other memory user) then the memory area cache should be cleaned after the memory update is finished by the CPU.
2. If the memory is updated by the peripheral (or DMA or any other memory user) and the CPU needs to get the result of such an update then the memory area cache should be invalidated prior to reading that area by the CPU. 
3. As you already mentioned the most conservative way is to use a cache flush that is clean+invalidate.

There is no harm in using cache flush, but from what I read it is that we want to erase the flash (actual memory update done by peripheral) and then read back the content to verify that page is actually erased. That is why I posted a comment.

Anyway let me browse the code and take a closer look.

Reply:
Thanks for the comment @pkarashchenko . I agree completely. I am a confused about who is the component (CPU or Flash Peripheral) that uses the Cache. In my first idea was something like this:

  1. The CPU performs a write operation into flash.
  2. Cache memory gets the new data and "retains" it.
  3. CPU performs a sector erase directly through the Flash Memory peripheral registers (Flash peripheral).
  4. Flash sector is erased.
  5. Cache data is flushed with delay (Synchronization failure).
  6. CPU reads Flash memory to check that is correctly erased. -> It finds non erased data. -> reports a failure.

But I agree that the process could be the following: 
  1. CPU performs a sector erase directly through the Flash Memory peripheral registers (Flash peripheral).
  2. Flash sector erased operation is managed though the Cache that "retains" the new data (erased data).
  3. CPU reads Flash memory to check that is correctly erased. -> It finds non erased data. -> reports a failure.
  4. Cache data is flushed with delay (Synchronization failure).

I discard the second one as I do not see that the erase operation in flash memory could involve directly the cache.

I think the second one is the approach you were talking about in your previous comment, am I right??

Reply:
The https://github.com/apache/incubator-nuttx/blob/ee82f7b5ca56a7f392843398e45f25773f8435d9/arch/arm/src/stm32h7/stm32h7b3xx_flash.c#L768 already contain similar fix, but there is an error in that line :)

I've looked in stm32h743753 reference manual (the "Embedded Flash memory" section) and see that flash read access is done via AXI interface over 256 bit read data buffers, so flash memory interface is "peripheral" in this case and erase command makes it to change the state on the flash. The CPU read of 256 bit read data buffers goes via cache (in other words data accessed from CPU on those flash addresses can be cached), so we need to invalidate the CPU data cache before next read access to those addresses after flash erase is done. In case of `stm32h7b3xx_flash.c` even if `up_invalidate_dcache()` is called before flash erase due to flash read access is protected by a semaphore the next flash read access is done in `stm32h7_israngeerased()`.

I think that the best place for `up_invalidate_dcache()` is after `ret = 0;` line in both `stm32h743xx` and `stm32h7b3xx`. If you want I can submit the change for both cases and you can verify it with your setup

Reply:
Yes please. Submit the changes and I test them in my setup. Thanks.

Reply:
Please try https://github.com/apache/incubator-nuttx/pull/7401